diff --git a/ply-image.c b/ply-image.c
index cfe91a9..1295751 100644
--- a/ply-image.c
+++ b/ply-image.c
@@ -436,6 +436,7 @@ main (int    argc,
   ply_image_t *image;
   ply_frame_buffer_t *buffer;
   int exit_code;
+  int scale_switch_fd;
 
   exit_code = 0;
 
@@ -471,6 +472,13 @@ main (int    argc,
 
   ply_image_free (image);
 
+  scale_switch_fd = open ("/sys/class/graphics/fb0/free_scale_switch", O_WRONLY);
+  if (scale_switch_fd >= 0)
+    {
+      write(scale_switch_fd, "0", 1);
+      close(scale_switch_fd);
+    }
+
   return exit_code;
 }
 
